crates/arroyo-df/src/plan.rs

```diff
@@ -171,43 +181,138 @@ impl TreeNodeRewriter for AggregateRewriter {
+        let mut visitor = WindowDetectingVisitor {
+            window: None,
+            fields: HashSet::new(),
+        };
         logical_plan.visit(&mut visitor)?;
         Ok(visitor.window.take())
     }
 }
 
+fn extract_column(expr: &Expr) -> Option<&Column> {
+    match expr {
+        Expr::Column(column) => Some(column),
+        Expr::Alias(Alias { expr, .. }) => extract_column(expr),
+        _ => None,
+    }
+}
+
 impl TreeNodeVisitor for WindowDetectingVisitor {
     type N = LogicalPlan;
 
-    fn pre_visit(&mut self, node: &Self::N) -> DFResult<VisitRecursion> {
+    fn post_visit(&mut self, node: &Self::N) -> DFResult<VisitRecursion> {
         match node {
+            LogicalPlan::Projection(projection) => {
+                let window_expressions = projection
+                    .expr
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(index, expr)| {
+                        if let Some(column) = extract_column(expr) {
+                            let input_field = projection
+                                .input
+                                .schema()
+                                .field_with_name(column.relation.as_ref(), &column.name);
+                            let input_field = match input_field {
+                                Ok(field) => field,
+                                Err(err) => {
+                                    return Some(Err(err));
+                                }
+                            };
+                            if self.fields.contains(input_field) {
+                                return self.window.clone().map(|window| Ok((index, window)));
+                            }
+                        }
+                        find_window(expr)
+                            .map(|option| option.map(|inner| (index, inner)))
+                            .map_err(|err| DataFusionError::Plan(err.to_string()))
+                            .transpose()
+                    })
+                    .collect::<DFResult<Vec<_>>>()?;
+                self.fields.clear();
+                for (index, window) in window_expressions {
+                    // if there's already a window they should match
+                    if let Some(existing_window) = &self.window {
+                        if *existing_window != window {
+                            return Err(DataFusionError::Plan(
+                                "window expressions do not match".to_string(),
+                            ));
+                        }
+                        self.fields.insert(projection.schema.field(index).clone());
+                    } else {
+                        // If the input doesn't have an input window, we shouldn't be creating a window.
+                        return Err(DataFusionError::Plan(
+                            "can't call window function in a select statement".to_string(),
```

Comment:
As a user, what does this mean?

Reply:
Gave it another shot. This is the error if you call one of our window UDFs without grouping by it.

Reply:
👍 that's much more clear